File: src/prod/src/managed/Api/src/System/Fabric/Description/ServiceHealthQueryDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{
    using System;
    using System.Fabric.Common;
    using System.Fabric.Health;
    using System.Fabric.Interop;
    using System.Globalization;
    using System.Text;

    /// <summary>
    /// <para>Provides query input for getting <see cref="System.Fabric.Health.ServiceHealth" />. Used by 
    /// <see cref="System.Fabric.FabricClient.HealthClient.GetServiceHealthAsync(System.Fabric.Description.ServiceHealthQueryDescription)" />.</para>
    /// </summary>
    public sealed class ServiceHealthQueryDescription
    {
        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Description.ServiceHealthQueryDescription" /> class.</para>
        /// </summary>
        /// <param name="serviceName">
        /// <para>The service name. Cannot be null.</para>
        /// </param>
        /// <exception cref="System.ArgumentNullException">
        /// <para>A null value was passed in for a required parameter.</para>
        /// </exception>
        public ServiceHealthQueryDescription(Uri serviceName)
        {
            Requires.Argument<Uri>("serviceName", serviceName).NotNull();
            this.ServiceName = serviceName;
        }

        /// <summary>
        /// <para>Gets the <see cref="System.Uri" /> of the service name.</para>
        /// </summary>
        /// <value>
        /// <para>The <see cref="System.Uri" /> of the service name.</para>
        /// </value>
        public Uri ServiceName
        {
            get;
            internal set;
        }

        /// <summary>
        /// <para>Gets or sets the <see cref="System.Fabric.Health.ApplicationHealthPolicy" /> used to evaluate health.</para>
        /// </summary>
        /// <value>
        /// <para>The <see cref="System.Fabric.Health.ApplicationHealthPolicy" /> used to evaluate health.</para>
        /// </value>
        /// <remarks>If not specified, the health store uses the application health policy of the parent application
        /// to evaluate the service health.</remarks>
        public ApplicationHealthPolicy HealthPolicy
        {
            get;
            set;
        }

        /// <summary>
        /// <para>Gets or sets the filter for the collection of <see cref="System.Fabric.Health.HealthEvent" /> reported on the service. Only events 
        /// that match the filter will be returned.</para>
        /// </summary>
        /// <value>
        /// <para>The <see cref="System.Fabric.Health.HealthEventsFilter" /> used to filter returned events.</para>
        /// </value>
        /// <remarks><para> Only events that match 
        /// the filter will be returned. All events will be used to evaluate the service aggregated health state.
        /// If the filter is not specified, all events are returned.</para></remarks>
        public HealthEventsFilter EventsFilter
        {
            get;
            set;
        }

        /// <summary>
        /// <para>Gets or sets the filter for <see cref="System.Fabric.Health.PartitionHealthState" /> children. Only children that match the filter 
        /// will be returned.</para>
        /// </summary>
        /// <value>
        /// <para>The <see cref="System.Fabric.Health.PartitionHealthStatesFilter" /> used to filter returned partition health states.</para>
        /// </value>
        /// <remarks><para> Only children that match the filter will be returned. 
        /// All children will be used to evaluate the service aggregated health state.
        /// If the filter is not specified, all partition children are returned.</para></remarks>
        public PartitionHealthStatesFilter PartitionsFilter
        {
            get;
            set;
        }

        /// <summary>
        /// Gets or sets the health statistics filter.
        /// </summary>
        /// <value>The health statistics filter.</value>
        /// <remarks>
        /// <para>
        /// The health statistics filter controls whether the <see cref="System.Fabric.Health.ServiceHealth"/> returned by the query
        /// contains the service health statistics. If not specified, the statistics are included.
        /// </para>
        /// </remarks>
        public ServiceHealthStatisticsFilter HealthStatisticsFilter
        {
            get;
            set;
        }

        /// <summary>
        /// Gets a string representation of the health query description.
        /// </summary>
        /// <returns>A string representation of the health query description.</returns>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append(string.Format(CultureInfo.CurrentCulture, "Service {0}", this.ServiceName));
            if (this.EventsFilter != null)
            {
                sb.Append(string.Format(CultureInfo.CurrentCulture, ", {0}", this.EventsFilter));
            }

            if (this.PartitionsFilter != null)
            {
                sb.Append(string.Format(CultureInfo.CurrentCulture, ", {0}", this.PartitionsFilter));
            }

            if (this.HealthPolicy != null)
            {
                sb.Append(string.Format(CultureInfo.CurrentCulture, ", {0}", this.HealthPolicy));
            }

            if (this.HealthStatisticsFilter != null)
            {
                sb.Append(string.Format(CultureInfo.CurrentCulture, ", {0}", this.HealthStatisticsFilter));
            }

            return sb.ToString();
        }

        internal IntPtr ToNative(PinCollection pinCollection)
        {
            var nativeServiceHealthQueryDescription = new NativeTypes.FABRIC_SERVICE_HEALTH_QUERY_DESCRIPTION();

            nativeServiceHealthQueryDescription.ServiceName = pinCollection.AddObject(this.ServiceName);
            if (this.HealthPolicy != null)
            {
                nativeServiceHealthQueryDescription.HealthPolicy = this.HealthPolicy.ToNative(pinCollection);
            }

            if (this.EventsFilter != null)
            {
                nativeServiceHealthQueryDescription.EventsFilter = this.EventsFilter.ToNative(pinCollection);
            }

            if (this.PartitionsFilter != null)
            {
                nativeServiceHealthQueryDescription.PartitionsFilter = this.PartitionsFilter.ToNative(pinCollection);
            }

            if (this.HealthStatisticsFilter != null)
            {
                var ex1 = new NativeTypes.FABRIC_SERVICE_HEALTH_QUERY_DESCRIPTION_EX1();
                ex1.HealthStatisticsFilter = this.HealthStatisticsFilter.ToNative(pinCollection);
                nativeServiceHealthQueryDescription.Reserved = pinCollection.AddBlittable(ex1);
            }

            return pinCollection.AddBlittable(nativeServiceHealthQueryDescription);
        }
    }
}

File: src/prod/src/ServiceModel/ServiceHealthQueryDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

StringLiteral const TraceSource("ServiceHealthQueryDescription");

ServiceHealthQueryDescription::ServiceHealthQueryDescription()
    : serviceName_()
    , healthPolicy_()
    , eventsFilter_()
    , partitionsFilter_()
    , healthStatsFilter_()
{
}

ServiceHealthQueryDescription::ServiceHealthQueryDescription(
    Common::NamingUri && serviceName,
    std::unique_ptr<ApplicationHealthPolicy> && healthPolicy)
    : serviceName_(move(serviceName))
    , healthPolicy_(move(healthPolicy))
    , eventsFilter_()
    , partitionsFilter_()
    , healthStatsFilter_()
{
}

ServiceHealthQueryDescription::~ServiceHealthQueryDescription()
{
}

ErrorCode ServiceHealthQueryDescription::FromPublicApi(FABRIC_SERVICE_HEALTH_QUERY_DESCRIPTION const & publicServiceHealthQueryDescription)
{
    // ServiceName
    {
        auto error = NamingUri::TryParse(
            publicServiceHealthQueryDescription.ServiceName,
            "ServiceName",
            serviceName_);
        if (!error.IsSuccess())
        {
            return error;
        }
    }

    // HealthPolicy
    if (publicServiceHealthQueryDescription.HealthPolicy != NULL)
    {
        ApplicationHealthPolicy appPolicy;
        auto error = appPolicy.FromPublicApi(*publicServiceHealthQueryDescription.HealthPolicy);
        if (!error.IsSuccess())
        {
            Trace.WriteInfo(TraceSource, "HealthPolicy::FromPublicApi error: {0}", error);
            return error;
        }

        healthPolicy_ = make_unique<ApplicationHealthPolicy>(move(appPolicy));
    }

    // EventsFilter
    if (publicServiceHealthQueryDescription.EventsFilter != NULL)
    {
        HealthEventsFilter eventsFilter;
        auto error = eventsFilter.FromPublicApi(*publicServiceHealthQueryDescription.EventsFilter);
        if (!error.IsSuccess())
        {
            Trace.WriteInfo(TraceSource, "EventsFilter::FromPublicApi error: {0}", error);
            return error;
        }

        eventsFilter_ = make_unique<HealthEventsFilter>(move(eventsFilter));
    }

    // PartitionsFilter
    if (publicServiceHealthQueryDescription.PartitionsFilter != NULL)
    {
        PartitionHealthStatesFilter partitionsFilter;
        auto error = partitionsFilter.FromPublicApi(*publicServiceHealthQueryDescription.PartitionsFilter);
        if (!error.IsSuccess())
        {
            Trace.WriteInfo(TraceSource, "PartitionsFilter::FromPublicApi error: {0}", error);
            return error;
        }

        partitionsFilter_ = make_unique<PartitionHealthStatesFilter>(move(partitionsFilter));
    }

    if (publicServiceHealthQueryDescription.Reserved != nullptr)
    {
        auto ex1 = static_cast<FABRIC_SERVICE_HEALTH_QUERY_DESCRIPTION_EX1*>(publicServiceHealthQueryDescription.Reserved);

        // ClusterHealthStatisticsFilter
        if (ex1->HealthStatisticsFilter != nullptr)
        {
            healthStatsFilter_ = make_unique<ServiceHealthStatisticsFilter>();
            auto error = healthStatsFilter_->FromPublicApi(*ex1->HealthStatisticsFilter);
            if (!error.IsSuccess())
            {
                Trace.WriteInfo(TraceSource, "HealthStatisticsFilter::FromPublicApi error: {0}", error);
                return error;
            }
        }
    }

    return ErrorCode::Success();
}

void ServiceHealthQueryDescription::SetQueryArguments(__in QueryArgumentMap & argMap) const
{
    argMap.Insert(Query::QueryResourceProperties::Service::ServiceName, serviceName_.Name);

    if (healthPolicy_)
    {
        argMap.Insert(
            Query::QueryResourceProperties::Health::ApplicationHealthPolicy,
            healthPolicy_->ToString());
    }

    if (eventsFilter_)
    {
        argMap.Insert(
            Query::QueryResourceProperties::Health::EventsFilter,
            eventsFilter_->ToString());
    }

    if (partitionsFilter_)
    {
        argMap.Insert(
            Query::QueryResourceProperties::Health::PartitionsFilter,
            partitionsFilter_->ToString());
    }

    if (healthStatsFilter_)
    {
        argMap.Insert(
            Query::QueryResourceProperties::Health::HealthStatsFilter,
            healthStatsFilter_->ToString());
    }
}
